### DISCUSSION TITLE
Add support for GPIO Switch and GPIO Binary sensor

### DISCUSSION THREAD

**Discussion by bbadrignans** [ffenix113/zigbee_home#43] [Ideas]
I think these 2 features would allow simple but useful applications for the project (such as driving relays, or using PIR and door sensors).
They should be inspired by : 
https://esphome.io/components/switch/gpio
https://esphome.io/components/binary_sensor/gpio

**Comment by bbadrignans** [comment]
Is it already done with on_off sensor and ias_zone ?

**Reply by ffenix113** [reply]
Switch is indeed supported with `on_off`, while binary sensor _could be_ substituted with `contact`, but it is not ideal, as it will be represented as IAS Zone indeed.
For binary sensor most appropriate cluster should be 3.14.5: binary input, which is not yet available.

Would you mind creating an issue for the binary input cluster?

Thank you.

**Reply by bbadrignans** [reply]
Thank you for your answer, great for these 2 features.
In my opinion, implement binary input is not a priority. 
IAS Zone is sufficient for now. Entity types can be customized in Homeassistant if one needs to show binary input or whatever.
But I can open an issue if you want.

I tried `on_off` but when I change state in homeassistant, the pin state does not change on the board (tried with a led).
The same with `contact`, changing electric state of the pin, does not change state in homeassistant.
I missed something or it is not yet functional ?

I noticed that turning on the `on_off` switch in Homeassistant generates this log then nothing, like a software crash : 

```
[00:00:00.000,579] <inf> ieee802154_nrf5: nRF5 802154 radio initialized
*** Booting nRF Connect SDK v3.5.99-ncs1 ***
[00:00:00.507,415] <inf> zigbee_app_utils: Production configuration is not present or invalid (status: -1)
[00:00:00.507,629] <inf> zigbee_app_utils: Zigbee stack initialized
[00:00:03.674,652] <inf> zigbee_app_utils: Joined network successfully on reboot signal (Extended PAN ID: abd034963397f935, PAN ID: 0x8991)
[00:00:10.554,840] <dbg> app: loop: ADC battery channel pin04: rated: 0, threshold: 0, current mv: 57599, percent(x2): 200
[00:01:10.620,758] <dbg> app: loop: ADC battery channel pin04: rated: 0, threshold: 0, current mv: 0, percent(x2): 200
[00:01:11.904,174] <inf> app: zcl_device_cb id 0
[00:01:11.904,296] <inf> app: zcl_device_cb s
```

Sample of my config : 
```
    - type: contact
      pin:
        port: 1
        pin: 10
    - type: on_off
      pin:
        port: 1
        pin: 9
```

**Reply by ffenix113** [reply]
For `on_off` the board should be set to router mode, as it need to be always ready to accept requests. Probably this can also be achieved with polling cluster, but I have not added/tried it yet.

As for `contact` - weird indeed. And log should also not break, for sure 🙂 . Could you please try to remove the directory in which source is generated and build firmware again, just to remove any possible left-overs in the code. Also can you please share full configuration so I can check it as well?

Sorry for late reply.

**Reply by bbadrignans** [reply]
No problem for the late reply, thanks for answer.

I tried the configuration bellow : just 1 `on_off` sensor as a router..
I use the unmodified dev branch. 
I cleared the build directory before.

I Tried also to change the name of the device (from FFenix113 to whatever), because I suppose there is a mess in Homeassistant / ZHA when a zigbee device changes is capabilities (I tried different combinations of sensors).

I used the button for factory reset (many times), and add/remove the device from Homeassistant.

Still the same in the logs, for instance bellow.
I tried here a factory reset, remove from ZHA, then peer again. It crashes as soon I turn on the switch in Homeassistant.
```
*** Booting nRF Connect SDK v3.5.99-ncs1 ***
[00:00:00.515,167] <inf> zigbee_app_utils: Production configuration is not present or invalid (status: -1)
[00:00:00.515,472] <inf> zigbee_app_utils: Zigbee stack initialized
[00:00:00.547,149] <inf> zigbee_app_utils: Unimplemented signal (signal: 54, status: 0)
[00:00:00.548,889] <inf> zigbee_app_utils: Joined network successfully on reboot signal (Extended PAN ID: abd034963397f935, PAN ID: 0x8991)
[00:00:23.713,134] <inf> zigbee_app_utils: Network left (leave type: 0)
[00:00:23.713,165] <inf> zigbee_app_utils: Started network rejoin procedure.
[00:00:28.616,851] <inf> zigbee_app_utils: Unimplemented signal (signal: 54, status: 0)
[00:00:29.333,099] <inf> zigbee_app_utils: Unimplemented signal (signal: 54, status: 0)
[00:00:29.334,594] <inf> zigbee_app_utils: Joined network successfully (Extended PAN ID: abd034963397f935, PAN ID: 0x8991)
[00:00:29.334,594] <inf> zigbee_app_utils: Network rejoin procedure stopped.
[00:00:52.343,017] <inf> app: zcl_device_cb id 0
[00:00:52.343,109] <inf> app: zcl_device_cb stat
```

If you find time to check this config, I'm interested by the result.
Maybe it is my toolchain setup or issue in ZHA.

My conf (without comments) : 
```
general:
  runevery: 1m
  board: nrf52840dongle_nrf52840
  
  ncs_toolchain_base: ~/ncs/
  zephyr_base: $ZEPHYR_BASE

  zigbee_channels: [11,13,15,16,17]

  flasher: nrfutil
  flasheroptions:
    port: /dev/ttyACM0

board:

  debug:
    enabled: true
    console: usb
    leds:
      enabled: true
      power: led0_green
  is_router: true
  i2c:
    - id: i2c0
      sda: 0.29
      scl:
        port: 0
        pin: 31
  leds:
    - id: led0_green

sensors:
    - type: on_off
      pin:
        port: 1
        pin: 9
```

**Reply by bbadrignans** [reply]
Logs from ZHA bellow. I named the manufacturer "zzzzz" instead of "FFenix113"  :

```
DEBUG (MainThread) [bellows.ezsp.protocol] Application frame received incomingMessageHandler: [<EmberIncomingMessageType.INCOMING_UNICAST: 0>, EmberApsFrame(profileId=260, clusterId=0, sourceEndpoint=1, destinationEndpoint=1, options=<EmberApsOption.APS_OPTION_RETRY|APS_OPTION_ENABLE_ROUTE_DISCOVERY|APS_OPTION_SOURCE_EUI64|APS_OPTION_DESTINATION_EUI64: 3392>, groupId=0, sequence=138), 232, -42, 0x6128, 255, 255, b'\x18\x07\x01\x04\x00\x00B\x06zzzzzz\x05\x00\x00B\nusb dongle']
DEBUG (MainThread) [bellows.zigbee.application] Received incomingMessageHandler frame with [<EmberIncomingMessageType.INCOMING_UNICAST: 0>, EmberApsFrame(profileId=260, clusterId=0, sourceEndpoint=1, destinationEndpoint=1, options=<EmberApsOption.APS_OPTION_RETRY|APS_OPTION_ENABLE_ROUTE_DISCOVERY|APS_OPTION_SOURCE_EUI64|APS_OPTION_DESTINATION_EUI64: 3392>, groupId=0, sequence=138), 232, -42, 0x6128, 255, 255, b'\x18\x07\x01\x04\x00\x00B\x06zzzzzz\x05\x00\x00B\nusb dongle']
DEBUG (MainThread) [zigpy.application] Received a packet: ZigbeePacket(timestamp=datetime.datetime(2024, 4, 21, 21, 39, 46, 677726, tzinfo=datetime.timezone.utc), src=AddrModeAddress(addr_mode=<AddrMode.NWK: 2>, address=0x6128), src_ep=1, dst=AddrModeAddress(addr_mode=<AddrMode.NWK: 2>, address=0x0000), dst_ep=1, source_route=None, extended_timeout=False, tsn=138, profile_id=260, cluster_id=0, data=Serialized[b'\x18\x07\x01\x04\x00\x00B\x06zzzzzz\x05\x00\x00B\nusb dongle'], tx_options=<TransmitOptions.NONE: 0>, radius=0, non_member_radius=0, lqi=232, rssi=-42)
DEBUG (MainThread) [zigpy.zcl] [0x6128:1:0x0000] Received ZCL frame: b'\x18\x07\x01\x04\x00\x00B\x06zzzzzz\x05\x00\x00B\nusb dongle'
DEBUG (MainThread) [zigpy.zcl] [0x6128:1:0x0000] Decoded ZCL frame header: ZCLHeader(frame_control=FrameControl<0x18>(frame_type=<FrameType.GLOBAL_COMMAND: 0>, is_manufacturer_specific=0, direction=<Direction.Server_to_Client: 1>, disable_default_response=1, reserved=0, *is_cluster=False, *is_general=True), tsn=7, command_id=1, *direction=<Direction.Server_to_Client: 1>)
DEBUG (MainThread) [zigpy.zcl] [0x6128:1:0x0000] Decoded ZCL frame: Basic:Read_Attributes_rsp(status_records=[ReadAttributeRecord(attrid=0x0004, status=<Status.SUCCESS: 0>, value=TypeValue(type=CharacterString, value='zzzzzz')), ReadAttributeRecord(attrid=0x0005, status=<Status.SUCCESS: 0>, value=TypeValue(type=CharacterString, value='usb dongle'))])
INFO (MainThread) [zigpy.device] [0x6128] Read model 'usb dongle' and manufacturer 'zzzzzz' from <Endpoint id=1 in=[basic:0x0000] out=[] status=<Status.ZDO_INIT: 1>>
INFO (MainThread) [zigpy.device] [0x6128] Discovered basic device information for <Device model='usb dongle' manuf='zzzzzz' nwk=0x6128 ieee=f4:ce:36:20:46:a2:61:55 is_initialized=True>
DEBUG (MainThread) [zigpy.application] Device is initialized <Device model='usb dongle' manuf='zzzzzz' nwk=0x6128 ieee=f4:ce:36:20:46:a2:61:55 is_initialized=True>
DEBUG (MainThread) [zigpy.quirks.registry] Checking quirks for zzzzzz usb dongle (f4:ce:36:20:46:a2:61:55)
DEBUG (MainThread) [zigpy.quirks.registry] Considering <class 'zhaquirks.xbee.xbee_io.XBeeSensor'>
DEBUG (MainThread) [zigpy.quirks] Fail because endpoint list mismatch: {232, 230} {1, 2, 242}
DEBUG (MainThread) [zigpy.quirks.registry] Considering <class 'zhaquirks.xbee.xbee3_io.XBee3Sensor'>
DEBUG (MainThread) [zigpy.quirks] Fail because endpoint list mismatch: {232, 230} {1, 2, 242}
DEBUG (MainThread) [zigpy.quirks.registry] Considering <class 'zhaquirks.tuya.ts0201.MoesTemperatureHumidtySensorWithScreen'>
DEBUG (MainThread) [zigpy.quirks] Fail because endpoint list mismatch: {1} {1, 2, 242}
DEBUG (MainThread) [zigpy.quirks.registry] Considering <class 'zhaquirks.smartthings.tag_v4.SmartThingsTagV4'>
DEBUG (MainThread) [zigpy.quirks] Fail because endpoint list mismatch: {1} {1, 2, 242}
DEBUG (MainThread) [zigpy.quirks.registry] Considering <class 'zhaquirks.smartthings.multi.SmartthingsMultiPurposeSensor'>
DEBUG (MainThread) [zigpy.quirks] Fail because endpoint list mismatch: {1} {1, 2, 242}
DEBUG (MainThread) [zigpy.quirks.registry] Considering <class 'zhaquirks.netvox.z308e3ed.Z308E3ED'>
DEBUG (MainThread) [zigpy.quirks] Fail because endpoint list mismatch: {1} {1, 2, 242}
DEBUG (MainThread) [zigpy.quirks.registry] Considering <class 'zhaquirks.gledopto.soposhgu10.SoposhGU10'>
DEBUG (MainThread) [zigpy.quirks] Fail because endpoint list mismatch: {11, 13} {1, 2, 242}
DEBUG (bellows.thread_0) [bellows.uart] Data frame: b'63b3b1a9cf2a40d3fbd26a1364a1a6ee967e'
DEBUG (bellows.thread_0) [bellows.uart] Sending: b'87009f7e'
DEBUG (MainThread) [homeassistant.components.zha.core.endpoint] Creating cluster handler for cluster id: 0 class: <class 'homeassistant.components.zha.core.cluster_handlers.general.BasicClusterHandler'>
DEBUG (MainThread) [homeassistant.components.zha.core.discovery] Discovering entities for endpoint: f4:ce:36:20:46:a2:61:55-1
DEBUG (MainThread) [homeassistant.components.zha.core.discovery] 'sensor' platform -> 'RSSISensor' using ['basic']
DEBUG (MainThread) [homeassistant.components.zha.core.discovery] 'sensor' platform -> 'LQISensor' using ['basic']
DEBUG (MainThread) [homeassistant.components.zha.core.endpoint] Creating cluster handler for cluster id: 6 class: <class 'homeassistant.components.zha.core.cluster_handlers.general.OnOffClusterHandler'>
DEBUG (MainThread) [homeassistant.components.zha.core.discovery] Discovering entities for endpoint: f4:ce:36:20:46:a2:61:55-2
DEBUG (MainThread) [homeassistant.components.zha.core.discovery] 'select' platform -> 'ZHAStartupOnOffSelectEntity' using ['on_off']
DEBUG (MainThread) [homeassistant.components.zha.core.discovery] Discovering entities for endpoint: f4:ce:36:20:46:a2:61:55-242
DEBUG (MainThread) [homeassistant.components.zha.core.device] [0x6128](usb dongle): starting availability checks - interval: 85
DEBUG (MainThread) [homeassistant.components.zha.core.discovery] Discovering entities for device: f4:ce:36:20:46:a2:61:55-zzzzzz usb dongle
DEBUG (MainThread) [homeassistant.components.zha.core.discovery] Attempting to discover quirks v2 entities for device: f4:ce:36:20:46:a2:61:55-zzzzzz usb dongle
DEBUG (MainThread) [homeassistant.components.zha.core.discovery] Device: f4:ce:36:20:46:a2:61:55-zzzzzz usb dongle is not a quirks v2 device - skipping discover_quirks_v2_entities
DEBUG (MainThread) [homeassistant.components.zha.core.gateway] device - 0x6128:f4:ce:36:20:46:a2:61:55 entering async_device_initialized - is_new_join: True
DEBUG (MainThread) [homeassistant.components.zha.core.gateway] device - 0x6128:f4:ce:36:20:46:a2:61:55 has joined the ZHA zigbee network
DEBUG (MainThread) [homeassistant.components.zha.core.device] [0x6128](usb dongle): started configuration
DEBUG (MainThread) [homeassistant.components.zha.core.cluster_handlers] [0x6128:ZDO](usb dongle): 'async_configure' stage succeeded
DEBUG (MainThread) [bellows.ezsp.protocol] Application frame received zigbeeKeyEstablishmentHandler: [f4:ce:36:20:46:a2:61:55, <EmberKeyStatus.TC_REQUESTER_VERIFY_KEY_SUCCESS: 52>]
DEBUG (MainThread) [bellows.zigbee.application] Received zigbeeKeyEstablishmentHandler frame with [f4:ce:36:20:46:a2:61:55, <EmberKeyStatus.TC_REQUESTER_VERIFY_KEY_SUCCESS: 52>]
DEBUG (MainThread) [zigpy.appdb] Error handling '_save_attribute' event with (f4:ce:36:20:46:a2:61:55, 1, 0, 4, 'zzzzzz', datetime.datetime(2024, 4, 21, 21, 39, 46, 681736, tzinfo=datetime.timezone.utc)) params: FOREIGN KEY constraint failed
DEBUG (MainThread) [homeassistant.components.zha.core.cluster_handlers] [0x6128:1:0x0000]: Configuring cluster attribute reporting
DEBUG (MainThread) [homeassistant.components.zha.core.cluster_handlers] [0x6128:1:0x0000]: finished cluster handler configuration
DEBUG (MainThread) [homeassistant.components.zha.core.cluster_handlers] [0x6128:2:0x0006]: Performing cluster binding
DEBUG (MainThread) [bellows.zigbee.application] Sending packet ZigbeePacket(timestamp=datetime.datetime(2024, 4, 21, 21, 39, 46, 716119, tzinfo=datetime.timezone.utc), src=AddrModeAddress(addr_mode=<AddrMode.NWK: 2>, address=0x0000), src_ep=0, dst=AddrModeAddress(addr_mode=<AddrMode.NWK: 2>, address=0x6128), dst_ep=0, source_route=None, extended_timeout=False, tsn=8, profile_id=0, cluster_id=<ZDOCmd.Bind_req: 0x0021>, data=Serialized[b'\x08Ua\xa2F 6\xce\xf4\x02\x06\x00\x03\xcf3\x00\xfe\xffD[8\x01'], tx_options=<TransmitOptions.NONE: 0>, radius=0, non_member_radius=0, lqi=None, rssi=None)
DEBUG (MainThread) [bellows.ezsp.protocol] Send command sendUnicast: (<EmberOutgoingMessageType.OUTGOING_DIRECT: 0>, 0x6128, EmberApsFrame(profileId=0, clusterId=33, sourceEndpoint=0, destinationEndpoint=0, options=<EmberApsOption.APS_OPTION_RETRY|APS_OPTION_ENABLE_ROUTE_DISCOVERY: 320>, groupId=0, sequence=8), 98, b'\x08Ua\xa2F 6\xce\xf4\x02\x06\x00\x03\xcf3\x00\xfe\xffD[8\x01')
DEBUG (MainThread) [homeassistant.components.zha.core.cluster_handlers] [0x6128:1:0x0000]: 'async_configure' stage succeeded
DEBUG (MainThread) [zigpy.appdb] Error handling '_save_attribute' event with (f4:ce:36:20:46:a2:61:55, 1, 0, 5, 'usb dongle', datetime.datetime(2024, 4, 21, 21, 39, 46, 681797, tzinfo=datetime.timezone.utc)) params: FOREIGN KEY constraint failed
```

**Reply by ffenix113** [reply]
Thank you! I was able to reproduce the issue, and will check it in detail in the evening.

**Reply by ffenix113** [reply]
I just [merged a fix](https://github.com/ffenix113/zigbee_home/pull/53), could you please check with latest commit?

Edit: FYI LEDs on the Dongle are active-low, so the pin that is used should also have `inverted: true`, like [here](https://github.com/ffenix113/zigbee_home/blob/5267ef5898b91f65a95f8ddfb3e221d1e6d11cf8/zigbee.yaml#L123-L128)

**Comment by bbadrignans** [comment]
@ffenix113  : Great, it works fine now. Thank you for this fix.

Some other points : 
- Would it be possible to add a name to the switch ? Currently having 2 ``on_off`` result in 2 switches with the same name in Homeassistant. Maybe by adding an optional name in the configuration like in esphome : 
```
   - type: on_off
     name : switch1
```
- Names would be interesting for other sensors too (`contact` or `ias_zone` for instance)
- It would also be great to configure for instance `DEVICE_INIT_BASIC_MODEL_ID` in yaml. It will be easier to differentiate multiple device in homeassistant
- I see in the doc that  `contact` and `ias_zone` "can work only with analog pins". Do you know why ? Because as I understand this sensor, it is just 0 or 1, it should work on any GPIO pin.
- I tried again `contact` and `ias_zone`. At boot the sensor is shown closed in homeassistant. If I put 3.3V on the pin, it turns opened but then I have a crash in zigbee_home with no log. Did you observe that ?

I added ias_zone like this : 
```
    - type: ias_zone
      pin:
        port: 0
        pin: 2
```

**Reply by ffenix113** [reply]
> Would it be possible to add a name to the switch ?

Unfortunately Zigbee does not provide interface for naming clusters. Though, possibly this could be implemented as a "vendor-specific" implementation of some other cluster, but otherwise - this is not part of the spec. I quickly glanced at Matter(i.e. next Zigbee) and it also does not seem to support setting names to clusters/attributes.

> It would also be great to configure for instance DEVICE_INIT_BASIC_MODEL_ID in yaml. It will be easier to differentiate multiple device in homeassistant

Fair point, and it was also something on my to-do list as well. I've just added this feature with 981058227b2a9a3a9176e5540fc12701d458b3fd. They can be set in `general` part of configuration:
```yaml
general:
  manufacturer: Someone
  device_name: Device
```

> I see in the doc that contact and ias_zone "can work only with analog pins". Do you know why ?

I just tried `ias_zone` with digital pin and it worked just the same. Will update docs to remove that note. Thank you!

> I tried again contact and ias_zone. At boot the sensor is shown closed in homeassistant. If I put 3.3V on the pin, it turns opened but then I have a crash in zigbee_home with no log. Did you observe that ?

I tried, and unfortunately I did see this behavior. Fixed with 492a2bc1d2d0ba2354a900a51b3f8e8c68596ed4. Thanks for reporting it!

**Reply by bbadrignans** [reply]
Ho I did not know for the absence of name in the specification. That would be very useful.
Annoying, but we still can make an alias in homeassistant. 

Thank you for manufacturer/device configuration : works perfectly.

For ias_zone, thank you for the doc update.
However, on my side, the fix did not work. I still have the same behaviour : 
- At startup state is close
- After I put 3.3V from VDDOUT pin, the device freeze : no log, on_off switch does not work
Do I missed something ?

**Reply by ffenix113** [reply]
> Annoying, but we still can make an alias in homeassistant.

I can understand this, and can agree. Probably it was designed for single-purpose devices, where this would not be as necessary.

> For ias_zone, thank you for the doc update.
However, on my side, the fix did not work. I still have the same behaviour

Hm, I will do some checks and push changes. Will notify when to try with new changes.

**Reply by ffenix113** [reply]
Sorry for late reply. I didn't do much changes, as I could not replicate the behavior on my Dongle.
Could you please update the binary and verify that the latest version is running by running `zigbee --version`? No options other then `--version`. It should print out the version and exit.
If it does - please to generate the firmware again.

**Comment by bbadrignans** [comment]
No problem for the delay, no hurry on my side.
Here is the version (last commit in develop branch : "fix: proper definition for device ctx with >4 endpoints") : 
`2024/05/11 22:48:10 main.go:79: go1.21.4, tag:(devel), version:[{vcs.modified true} {vcs.revision 62e0b104595625fdce74b710da1f9d722e48c58b}]`

I figure out that it works fine when having 1 or 2 on_off and 1 ias_zone.
But doesn't with 1 ias_zone and 1 dht sensor (or 1 bme680 sensor).
Maybe you can reproduce with this information.

Regards

**Reply by ffenix113** [reply]
Thank you for understanding. I am working and studying at the same time, which takes quite a few hours from the week. But this project is  still active, and even more so - I am designing some hardware for it. Just to be clear that I am not just ignoring the issues/discussions 🙂 

Thank you for the info, I can definitely reproduce that and will do so ASAP.

**Reply by ffenix113** [reply]
Hello @bbadrignans,

Thank you for the wait. I think I have fixed the issue that you had, so if you would have ability - please try with [latest release](https://github.com/ffenix113/zigbee_home/releases/tag/0.3).

For now I will close the discussion, but if you would see that you have any more issues - please feel free to open the issue, or a discussion for questions!
